File: fabric/src/main/java/juuxel/adorn/platform/fabric/MenuBridgeImpl.java
package juuxel.adorn.platform.fabric;

import juuxel.adorn.platform.MenuBridge;
import juuxel.adorn.util.Logging;
import net.fabricmc.fabric.api.menu.v1.ExtendedMenuProvider;
import net.fabricmc.fabric.api.menu.v1.ExtendedMenuType;
import net.minecraft.core.BlockPos;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.chat.Component;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.MenuProvider;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.MenuType;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;

public final class MenuBridgeImpl implements MenuBridge {
    public static final MenuBridgeImpl INSTANCE = new MenuBridgeImpl();
    private static final Logger LOGGER = Logging.logger();

    @Override
    public void open(Player player, @Nullable MenuProvider factory, BlockPos pos) {
        if (factory == null) {
            LOGGER.warn("[Adorn] Menu factory is null, please report this!", new Throwable("Stacktrace").fillInStackTrace());
            return;
        }

        if (!player.level().isClientSide()) {
            // ^ technically not needed as vanilla safeguards against it,
            // but no need to create the extra factory on the client

            player.openMenu(new ExtendedMenuProvider<>() {
                @Override
                public @Nullable AbstractContainerMenu createMenu(int syncId, Inventory playerInventory, Player player) {
                    return factory.createMenu(syncId, playerInventory, player);
                }

                @Override
                public Component getDisplayName() {
                    return factory.getDisplayName();
                }

                @Override
                public BlockPos getScreenOpeningData(ServerPlayer player) {
                    return pos;
                }
            });
        }
    }

    @Override
    public <M extends AbstractContainerMenu, D> MenuType<M> createType(Factory<M, D> factory, StreamCodec<? super RegistryFriendlyByteBuf, D> packetCodec) {
        return new ExtendedMenuType<>(factory::create, packetCodec);
    }
}

File: fabric/src/main/kotlin/juuxel/adorn/platform/fabric/MenuBridgeImpl.kt
package juuxel.adorn.platform.fabric

import juuxel.adorn.platform.MenuBridge
import juuxel.adorn.util.logger
import net.fabricmc.fabric.api.screenhandler.v1.ExtendedScreenHandlerFactory
import net.fabricmc.fabric.api.screenhandler.v1.ExtendedScreenHandlerType
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.menu.Menu
import net.minecraft.menu.MenuType
import net.minecraft.menu.NamedMenuFactory
import net.minecraft.network.PacketByteBuf
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.math.BlockPos

object MenuBridgeImpl : MenuBridge {
    private val LOGGER = logger()

    override fun open(player: PlayerEntity, factory: NamedMenuFactory?, pos: BlockPos) {
        if (factory == null) {
            LOGGER.warn("[Adorn] Menu factory is null, please report this!", Throwable("Stacktrace").fillInStackTrace())
            return
        }

        if (!player.world.isClient) {
            // ^ technically not needed as vanilla safeguards against it,
            // but no need to create the extra factory on the client
            player.openMenu(
                object : ExtendedScreenHandlerFactory, NamedMenuFactory by factory {
                    override fun writeScreenOpeningData(player: ServerPlayerEntity, buf: PacketByteBuf) {
                        buf.writeBlockPos(pos)
                    }
                }
            )
        }
    }

    override fun <M : Menu> createType(factory: (syncId: Int, inventory: PlayerInventory, buf: PacketByteBuf) -> M): MenuType<M> =
        ExtendedScreenHandlerType(factory)
}
